concept3.py

```diff
@@ -0,0 +1,108 @@
+import sys
+from prompt_toolkit import PromptSession
+import signal
+import asyncio
+
+
+class MentatInterrupt(Exception):
+    pass
+
+
+class MentatInterface:
+    # should be implemented by subclasses
+    def display(self, content, color=None, end="\n"):
+        pass
+
+    # should be implemented by subclasses
+    async def get_user_input(self, prompt=None, options=None):
+        pass
+
+    # convenience function
+    async def ask_yes_no(self, prompt=None):
+        return (await self.get_user_input(prompt, options=["y", "n"])).lower() == "y"
+
+
+class TerminalMentatInterface(MentatInterface):
+    """Used to communicate with the terminal"""
+
+    def __init__(self):
+        self.session = PromptSession()
+
+    def display(self, content, color=None, end="\n"):
+        print(content, end=end, flush=True)
+        global sigint
+        if sigint:
+            sigint = False
+            raise MentatInterrupt()
```

Comment:
the idea here was we'd check the global whenever Mentat tries to display something, and raise a `MentatInterrupt` if the user had pressed ctrl-c (or, for the VSCode extension, if the user had pressed some button or hotkey to interrupt / quit).

The `MentatInterrupt` gets caught in a try/except wrapped around streaming, to just interrupt the stream. If we aren't in streaming, it gets caught higher up to quit Mentat.